lib/elixir/src/elixir_expand.erl

```diff
@@ -814,11 +814,20 @@ expand_remote(Receiver, DotMeta, Right, Meta, Args, S, SL, #{context := Context}
 
   case {Context, lists:keyfind(no_parens, 1, Meta)} of
     {guard, {no_parens, true}} when is_tuple(Receiver) ->
-      {{{'.', DotMeta, [Receiver, Right]}, Meta, []}, SL, E};
+      case S#elixir_ex.bitsize of
+        true ->
+          form_error(Meta, E, ?MODULE, {invalid_map_lookup_bitsize, Receiver, Right});
```

Comment:
This specific case (map lookup) seems to be a limitation on the OTP side.. or maybe extra work is needed to prevent an Erlang compiler error.

I just decided to throw an error in this edge case. In any case, it is possible to work on it in a separate PR.

Reply:
Hrm, which error do you get? This works for me:

```erlang
-module(foo).
-compile(export_all).

foo(Map) ->
  <<7:(map_get(foo, Map))>> = <<255:3>>.
```

Then called with:

```
foo:foo(#{foo => 3}).
```

Reply:
You're totally right, it's not a limitation in Erlang. My bad!

This is the error I get if I don't add the if clause:

```
iex(1): foo = %{ba: 3}
%{ba: 3}
iex(2)> <<ff::size(foo.ba)>> = <<1,2,3>>
** (ErlangError) Erlang error: {:illegal_pattern, {:bin, 2, [{:bin_element, 2, {:var, 2, :_ff@1}, {:case, [generated: true, location: 2], {:var, 2, :_@0}, [{:clause, [generated: true, location: 2], [{:map, 2, [{:map_field_exact, 2, {:atom, 2, :ba}, {:var, [generated: true, location: 2], :_@2}}]}], [], [{:var, [generated: true, location: 2], :_@2}]}, {:clause, [generated: true, location: 2], [{:var, [generated: true, location: 2], :_@2}], [[{:call, [generated: true, location: 2], {:remote, [generated: true, location: 2], {:atom, [generated: true, location: 2], :erlang}, {:atom, [generated: true, location: 2], :is_map}}, [{:var, [generated: true, location: 2], :_@2}]}]], [{:call, 2, {:remote, 2, {:atom, 2, :erlang}, {:atom, 2, :error}}, [{:tuple, 2, [{:atom, 2, :badkey}, {:atom, 2, :ba}, {:var, [generated: true, location: 2], :_@2}]}]}]}, {:clause, [generated: true, location: 2], [{:var, [generated: true, location: 2], :_@2}], [], [{:call, [generated: true, location: 2], {:remote, [generated: true, location: 2], {:var, [generated: true, location: 2], :_@2}, {:atom, 2, :ba}}, []}]}]}, [:integer]}]}}
```

It looks like the `foo.bar` is being expanded to a case block, but this is something I understand it is not supposed to happen when the context is `guard` (https://github.com/elixir-lang/elixir/blob/main/lib/elixir/src/elixir_erl_pass.erl#L220-L250).

I guess I need to check and learn more about this case to properly cover this case. I can try it 👍

Reply:
@jmbejar oh, that's an easy fix! In here:

https://github.com/elixir-lang/elixir/blob/main/lib/elixir/src/elixir_erl_pass.erl#L542

You have to do the same dance with S, something like:

    S#elixir_erl{context=guard}

And then revert it back!

Can you please give it a try? Let me know if you have questions!

Reply:
Adding a test around this would be enough IMO!

Reply:
Gotcha.. I was around that part of the code already, but you are giving me the proper solution 👍. I think it should be enough.

Reply:
Confirmed that works 👍